[Documentation Code Camp] docstrings for emitted signals

meguiraun opened the pull request:
Initial work for documenting emitted signals. I initially focused  on the abstract classes (not all done yet 🤪)

Example output:
<img width="781" alt="image" src="https://github.com/mxcube/mxcubecore/assets/12758327/a0808761-8ebf-4324-aac9-bb6ad5c104a0">

meguiraun committed:
docstrings for emitted signals

fabcor-maxiv commented:
Wouldn't it make more sense to document the emitted signals at the class level rather than at the method level?

If I use an *actuator* I want to know what signals it emits and when (under what circumstances, i.e. what is the meaning of the signal), but I do not need to know which of the actuator's methods actually emits the signal, do I?

And same thing, if I want to implement a new (concrete) actuator, I want to know what signals my actuator should emit, but I will choose which methods will actually emit those signals.

cc @elmjag

beteva commented:
@meguiraun thanks for the work. Good start. There is one slight problem - I think AbstractAperture is an obsolete class, There will be very soon a new PR to handle aperture test and mock up, but they inherit from AbstractNState. In this context, would you mind if we see together which are the obsolete abstract classes.
@fabcor-maxiv I guess it is all right if we document the signals on the class level only. Actually this is the case right now, but certainly needs a bit of explanation which signal means what.

Maybe we should make a general signals documentation file, which lists all the "standard" signals, like valueChanged, limitsChanges, stateChanged, specificStateChanged, as well as the methods defined in  BaseHardwareObjects.py - the emit method (namely its signature in order to explain how to emit a signal) and the force_emit_signals method,  available for overloading . If we go for this, than we could explain the usage of these signals and in the classes only mention which are the emitted  signals,  and only document if there are other signals, specific for the class. What do you think?

rhfogh commented:
I agree with @beteva - The biggest need is what are the general kinds of signals that are used to tie the application together, where they come from, and what they do. In a way the biggest question is not so much 'which signals does this class emit' but 'which signals should I listen for' and 'which signals do I need to emit in my own class?'.

meguiraun commented:
okey, I see your point. I will leave the methods out.

yes, totally agree with a general documentation page (why, where in code are coming, some basic example, etc.), I had it in mind but I thought to start with documenting the code

meguiraun committed:
draft for signal documentation

meguiraun committed:
removing method doctrings for emitted signals

meguiraun commented:
1. Added a new md with signal documentation, very early, please @beteva @rhfogh  have a look, and if you feel so commit any improvement
2. Updated, let me know if the "new" AbstracActuator class documentation is ok:

<img width="688" alt="image" src="https://github.com/mxcube/mxcubecore/assets/12758327/26fd8081-6f16-44f5-8861-0567efd016c1">

fabcor-maxiv commented on mxcubecore/HardwareObjects/abstract/AbstractActuator.py:
```diff
@@ -37,7 +44,25 @@
+        read only flag, specified by xml property, otherwise False
+
+    username : int
+        the number of legs the animal has (default 4)
```
What!? :D

```suggestion
```

fabcor-maxiv commented on mxcubecore/HardwareObjects/abstract/AbstractActuator.py:
```diff
@@ -37,7 +44,25 @@
-    """Abstract actuator"""
+    """Abstract actuator
+
+    Attributes
+    ----------
+    _nominal_value : float
+        Current actuator value
+    default_value : float
+        value specified by xml property, otherwise None
+    _nominal_limits : tuple
+        values specified by xml property, otherwise None
+    actuator_name : str
+        actuator name specified by xml property, otherwise None
+    read_only : bool
+        read only flag, specified by xml property, otherwise False
+
```
```suggestion
    Attributes:
        _nominal_value (float):
            Current actuator value
        default_value (float):
            value specified by XML property, otherwise `None`
        _nominal_limits (tuple):
            values specified by XML property, otherwise `None`
        actuator_name (str):
            actuator name specified by XML property, otherwise `None`
        read_only (bool):
            read only flag, specified by XML property, otherwise `False`
```

fabcor-maxiv commented on mxcubecore/HardwareObjects/abstract/AbstractBeam.py:
```diff
@@ -213,6 +219,13 @@ def evaluate_beam_info(self):
         Reemits beamSizeChanged and beamInfoChanged signals
+
+        Emits:
+            beamSizeChanged:  ('beamSizeChanged', (width, height)))
```
```suggestion
            beamSizeChanged:  ('beamSizeChanged', (width, height))
```

fabcor-maxiv commented on docs/source/dev/hwobj_signals.md:
Personal preference, we do not have a policy for this as far as I know: I tend to prefer "_[Semantic Line Breaks](https://sembr.org/)_".

rhfogh replied:
Fabien only just introduced me to semantic line breaks, but I am heartily in favour.

elmjag replied:
wait, semantic break is the same things as putting each sentence on it's own line?

fabcor-maxiv replied:
> semantic break is the same things as putting each sentence on it's own line?

No.

I never read the full document, only the introduction, I do not see much point in being strict about this. But to me, the important thing is to break long paragraphs where it makes sense logically.

So in Markdown I would write like this:

```none
I never read the full document, only the introduction,
I do not see much point in being strict about this.
But to me, the important thing is to break long paragraphs
where it makes sense logically.
```

fabcor-maxiv commented on docs/source/dev/hwobj_signals.md:
```diff
@@ -0,0 +1,208 @@
+
+:Created: 20240318
+
+MXCuBE relies heavily on signals beeing emmited and listened to by many elements. For example, a hardware object may listen to other lower level hardware objects in order to update values after some calculation. But it is also critical in the UI (both web and QT), in both cases they expect periodic signal updates for displaying the most recent information to the user (e.g. motor positions, data c ...
```
```suggestion
MXCuBE relies heavily on signals being emitted and listened to by many elements. For example, a hardware object may listen to other lower level hardware objects in order to update values after some calculation. But it is also critical in the UI (both web and Qt), in both cases they expect periodic signal updates for displaying the most recent information to the user (e.g. motor positions, data collection state, etc.)
```

fabcor-maxiv commented on docs/source/dev/hwobj_signals.md:
```diff
@@ -0,0 +1,208 @@
+
+## Implementation
+
+Depending on the installed modules signals are emitted using [Louie](https://pypi.org/project/Louie/) or [PyDispatcher](https://pypi.org/project/PyDispatcher/). The former being based on the later. The developer does not need to deal with the differences between those two modules as it is already being handled in the file [dispatcher](https://github.com/mxcube/mxcubecore/blob/develop/mxcubecore/d ...
```
```suggestion
Depending on the installed modules, signals are emitted using [Louie](https://pypi.org/project/Louie/) or [PyDispatcher](https://pypi.org/project/PyDispatcher/). The former being based on the later. The developer does not need to deal with the differences between those two modules as it is already being handled in the file [dispatcher](https://github.com/mxcube/mxcubecore/blob/develop/mxcubecore/dispatcher.py).
```

fabcor-maxiv commented on docs/source/dev/hwobj_signals.md:
```diff
@@ -0,0 +1,208 @@
+## Implementation
+
+Depending on the installed modules signals are emitted using [Louie](https://pypi.org/project/Louie/) or [PyDispatcher](https://pypi.org/project/PyDispatcher/). The former being based on the later. The developer does not need to deal with the differences between those two modules as it is already being handled in the file [dispatcher](https://github.com/mxcube/mxcubecore/blob/develop/mxcubecore/d ...
+ **_NOTE:_**  can we remove of of those dependencies?
```
```suggestion
 **_NOTE:_**  can we remove any of those dependencies?
```

fabcor-maxiv commented on docs/source/dev/hwobj_signals.md:
```diff
@@ -0,0 +1,208 @@
+
+When certain events or conditions occur within a hardware object, corresponding signals are emitted to inform connected components or modules about these changes.
+
+The [BaseHardwareObject](https://github.com/mxcube/mxcubecore/blob/develop/mxcubecore/BaseHardwareObjects.py) class serves as the base class for all hardware objects in MXCuBE. It includes methods for defining and emitting signals, allowing derived classes to customize signal emission based on their specific requirements.
```
```suggestion
The {py:class}`mxcubecore.BaseHardwareObjects.HardwareObject` class serves as the base class for all hardware objects in MXCuBE. It includes methods for defining and emitting signals, allowing derived classes to customize signal emission based on their specific requirements.
```

fabcor-maxiv commented on docs/source/dev/hwobj_signals.md:
```diff
@@ -0,0 +1,208 @@
+
+Signals are typically emitted when the state of a hardware object changes, such as when it becomes ready for operation, encounters an error, or completes a task. Additionally, signals may be emitted to indicate changes in parameters or settings of the hardware, such as new setpoints, values, or configuration options.
+
+To emit a signal, derived classes can use the `emit` method provided by the BaseHardwareObject class. This method takes the name of the signal as an argument and optionally includes additional data or parameters to pass along with the signal. This method calls the `dispatcher.send` method.
```
```suggestion
To emit a signal, derived classes can use the {py:meth}`mxcubecore.BaseHardwareObjects.HardwareObjectMixin.emit` method provided by the {py:class}`HardwareObject` class. This method takes the name of the signal as an argument and optionally includes additional data or parameters to pass along with the signal. This method calls the `dispatcher.send` method.
```

fabcor-maxiv commented on docs/source/dev/hwobj_signals.md:
```diff
@@ -0,0 +1,208 @@
+To emit a signal, derived classes can use the `emit` method provided by the BaseHardwareObject class. This method takes the name of the signal as an argument and optionally includes additional data or parameters to pass along with the signal. This method calls the `dispatcher.send` method.
+
+
+From the _BaseHardwareObject_ class (removing extra lines for brevity):
```
```suggestion
From the {py:class}`HardwareObject` class (removing extra lines for brevity):
```

fabcor-maxiv commented on docs/source/dev/hwobj_signals.md:
````diff
@@ -0,0 +1,208 @@
+        dispatcher.send(signal, self, *args)
+```
+
+So, in a custom hardware object, since inhnerits from _BaseHardwareObject_, one only needs to call: 
````
```suggestion
So, in a custom hardware object, since it inherits from {py:class}`HardwareObject`, one only needs to call: 
```

fabcor-maxiv commented on docs/source/dev/hwobj_signals.md:
````diff
@@ -0,0 +1,208 @@
+
+### Receive
+
+_BaseHardwareObject_ implements the following ```connect```method, built around the homonymous method of _PyDispatcher_. Making it more convenient to use. The functions provides syntactic sugar ; Instead of ```self.connect(self, "signal", slot)``` it is possible to do ```self.connect("signal", slot)```
````
```suggestion
{py:class}`HardwareObject` implements the following {py:meth}`HardwareObjectMixin.connect` method, built around the homonymous method of _PyDispatcher_. Making it more convenient to use. The functions provides syntactic sugar: instead of `self.connect(self, "signal", slot)` it is possible to do `self.connect("signal", slot)`.
```

fabcor-maxiv commented on docs/source/dev/hwobj_signals.md:
````diff
@@ -0,0 +1,208 @@
+
+_BaseHardwareObject_ implements the following ```connect```method, built around the homonymous method of _PyDispatcher_. Making it more convenient to use. The functions provides syntactic sugar ; Instead of ```self.connect(self, "signal", slot)``` it is possible to do ```self.connect("signal", slot)```
+
+From the [BaseHardwareObject](https://github.com/mxcube/mxcubecore/blob/develop/mxcubecore/BaseHardwareObjects.py#L878) class (removing extra lines for brevity):
````
```suggestion
From the {py:meth}`HardwareObjectMixin.connect` method (removing extra lines for brevity):
```

fabcor-maxiv commented on docs/source/dev/hwobj_signals.md:
````diff
@@ -0,0 +1,208 @@
+    self.connect(some_other_hwobj, "a_signal", callback_method)
+```
+
+This assumes that ```some_other_hwobj``` is linked in the custom hwobj initialization, and _callback_method_ must exist, otherwise an exception will happen once the signal is received.
````
```suggestion
This assumes that `some_other_hwobj` is linked in the custom hardware object initialization, and `callback_method` must exist, otherwise an exception will happen once the signal is received.
```

fabcor-maxiv commented on docs/source/dev/hwobj_signals.md:
````diff
@@ -0,0 +1,208 @@
+
+This assumes that ```some_other_hwobj``` is linked in the custom hwobj initialization, and _callback_method_ must exist, otherwise an exception will happen once the signal is received.
+
+If the sender hwobj has a method named ```connect_notify```, it will be called on connect. Since this connect happens at application init, this typically triggers the emission of all signals during initialization, and thus all receivers start with the most recent value.
````
```suggestion
If the sender hardware object has a method named `connect_notify`, it will be called on connect. Since this connect happens at application initialization, this typically triggers the emission of all signals during initialization, and thus all receivers start with the most recent value.
```

fabcor-maxiv commented on docs/source/dev/hwobj_signals.md:
```diff
@@ -0,0 +1,208 @@
+
+As you can see, the second hardware object receives and processes first one's signal.
+
+> At least one entry must appear in the beamline config yml file, in this case I left the procedure mockup, all the other mockups are commented. Thats why only a few items appear in the loading table.
```
```suggestion
> At least one entry must appear in the beamline's YAML configuration file. In this case I left the procedure mockup only, all the other mockups are commented. That is why only a few items appear in the loading table.
```

fabcor-maxiv reviewed (changes requested):
I do not think there is a `BaseHardwareObject` class. So I replaced with `HardwareObject`. But in most cases what we talk about is actually in the `HardwareObjectMixin`.

I tried the Sphinx/docutils cross-reference roles, we'll see if we get better links.

In principle we should be able to configure Sphinx to keep a copy of the source code, so that we can have links to source code in the documentation (not links to GitHub).

And in general, not only for (Sphinx) documentation, we should avoid links to a branch, but use permalinks that link to a specific commit.

fabcor-maxiv commented:
In many of the modified docstrings, it feels like adding the correct type hints would be nearly as efficient (especially for the return values).

rhfogh commented:
@fabcor-maxiv  HardwareObjectMixin woudl be the right choice throughout. HardwareObjects inherit from *either* HardwareObject *or* HardwareObjectYaml, but both inherit from HardwareObjectMixin,.

meguiraun committed:
Update mxcubecore/HardwareObjects/abstract/AbstractActuator.py

Co-authored-by: fabcor <fabien.coronis@maxiv.lu.se>

meguiraun committed:
Update mxcubecore/HardwareObjects/abstract/AbstractBeam.py

Co-authored-by: fabcor <fabien.coronis@maxiv.lu.se>

meguiraun committed:
Update docs/source/dev/hwobj_signals.md

Co-authored-by: fabcor <fabien.coronis@maxiv.lu.se>

Update docs/source/dev/hwobj_signals.md

Co-authored-by: fabcor <fabien.coronis@maxiv.lu.se>

Update docs/source/dev/hwobj_signals.md

Co-authored-by: fabcor <fabien.coronis@maxiv.lu.se>

Update docs/source/dev/hwobj_signals.md

Co-authored-by: fabcor <fabien.coronis@maxiv.lu.se>

Update docs/source/dev/hwobj_signals.md

Co-authored-by: fabcor <fabien.coronis@maxiv.lu.se>

Update docs/source/dev/hwobj_signals.md

Co-authored-by: fabcor <fabien.coronis@maxiv.lu.se>

Update docs/source/dev/hwobj_signals.md

Co-authored-by: fabcor <fabien.coronis@maxiv.lu.se>

Update docs/source/dev/hwobj_signals.md

Co-authored-by: fabcor <fabien.coronis@maxiv.lu.se>

Update docs/source/dev/hwobj_signals.md

Co-authored-by: fabcor <fabien.coronis@maxiv.lu.se>

meguiraun committed:
remove wrong line

meguiraun committed:
Update docs/source/dev/hwobj_signals.md

Co-authored-by: fabcor <fabien.coronis@maxiv.lu.se>

meguiraun committed:
properly cite HardwareObjecMixin

meguiraun committed:
initial table with signals

meguiraun commented:
preview in https://mxcubecore--866.org.readthedocs.build/en/866/dev/hwobj_signals.html

beteva commented:
Very good job @meguiraun 👍
The device and the equipment signals are definitely deprecated. Quite some time ago there was a decision not to use Device and Equipment class any more.
You are right, we should remove the update signal and only use valueChanged.
You can also remove the xrf related signals - in my WIP for xrf I've changed them all to stateChanged

marcus-oscarsson commented:
Well done ! very nice !. As Antonia already pointed out, we should really start to remove the Device and Equipment classes.

rhfogh commented on docs/source/dev/hwobj_signals.md:
```diff
@@ -0,0 +1,244 @@
+
+When certain events or conditions occur within a hardware object, corresponding signals are emitted to inform connected components or modules about these changes.
+
+The {py:class}`mxcubecore.BaseHardwareObjects.HardwareObject` class serves as the base class for all hardware objects in MXCuBE. It includes methods for defining and emitting signals, allowing derived classes to customize signal emission based on their specific requirements.
```
Strictly speaking it is the HardwareObject  OR HardwareObjectYaml class (both inherit from HardwareObjectMixin). Once we unify the YAML and XML configuration this distinction should hopefully disappear.

meguiraun replied:
added this clarification in the text

meguiraun committed:
mark deprecated signals

meguiraun committed:
queue and diff

meguiraun commented:
while the signals are still there, I simply marked as deprecated. 
Also added queue and diffractometer

meguiraun committed:
some data types; typos

axelboc commented on docs/source/dev/hwobj_signals.md:
```diff
@@ -0,0 +1,339 @@
+
+## Implementation
+
+Depending on the installed modules, signals are emitted using [Louie](https://pypi.org/project/Louie/) or [PyDispatcher](https://pypi.org/project/PyDispatcher/). The former being based on the later. The developer does not need to deal with the differences between those two modules as it is already being handled in the file [dispatcher](https://github.com/mxcube/mxcubecore/blob/develop/mxcubecore/ ...
```
```suggestion
Depending on the installed modules, signals are emitted using [Louie](https://pypi.org/project/Louie/) or [PyDispatcher](https://pypi.org/project/PyDispatcher/). The former being based on the latter. The developer does not need to deal with the differences between those two modules as it is already being handled in the file [`dispatcher.py`](https://github.com/mxcube/mxcubecore/blob/develop/mxcubecore/dispatcher.py).
```

axelboc commented on docs/source/dev/hwobj_signals.md:
```diff
@@ -0,0 +1,339 @@
+
+The {py:class}`mxcubecore.BaseHardwareObjects.HardwareObject` class serves as the base class for all hardware objects in MXCuBE. It includes methods for defining and emitting signals, allowing derived classes to customize signal emission based on their specific requirements.
+
+>Strictly speaking it is the HardwareObject OR HardwareObjectYaml class (both inherit from HardwareObjectMixin). Once we unify the YAML and XML configuration this distinction should hopefully disappear.
```
```suggestion
> Strictly speaking it is the `HardwareObject` OR `HardwareObjectYaml` class (both inherit from `HardwareObjectMixin`). Once we unify the YAML and XML configuration, this distinction should hopefully disappear.
```

axelboc commented on docs/source/dev/hwobj_signals.md:
````diff
@@ -0,0 +1,339 @@
+To emit a signal, derived classes can use the {py:meth}`mxcubecore.BaseHardwareObjects.HardwareObjectMixin.emit` method provided by the {py:class}`HardwareObject` class. This method takes the name of the signal as an argument and optionally includes additional data or parameters to pass along with the signal. This method calls the `dispatcher.send` method.
+
+
+From the {py:class}`HardwareObjectMixin` class (removing extra lines for brevity):
+
+```
+    def emit(self, signal: Union[str, object, Any], *args) -> None:
+        signal = str(signal)
+
+        if len(args) == 1:
+            if isinstance(args[0], tuple):
+                args = args[0]
+        dispatcher.send(signal, self, *args)
+```
````
Not convinced this code sample is needed. Same with `connect` below. One can always go look in the classes mentioned. There is also the risk that the docs might fall out of sync in the future if the code ever changes.

meguiraun replied:
yeah I know, but I still think it makes the description easier to understand. That code has not changed much anyway in the recent mxcube history

axelboc commented on docs/source/dev/hwobj_signals.md:
```diff
@@ -0,0 +1,339 @@
+
+As you can see, the second hardware object receives and processes first one's signal.
+
+> At least one entry must appear in the beamline's YAML configuration file. In this case I left the procedure mockup only, all the other mockups are commented. That is why only a few items appear in the loading table.
```
Maybe reword to avoid using the first person "I"?

axelboc commented on docs/source/dev/hwobj_signals.md:
```diff
@@ -0,0 +1,339 @@
+
+## General signals List
+
+The following tables list the generic signals as well the signals from the some of the most important hardware objects. 
```
```suggestion
The following tables list the generic signals as well as the signals from some of the most important hardware objects. 
```

axelboc commented on mxcubecore/HardwareObjects/abstract/AbstractActuator.py:
```diff
@@ -22,7 +22,14 @@
-Emits signals valueChanged and limitsChanged.
+
+Emits:
+    valueChanged: during hwobj initialization and when setting a new value. Format: (   valueChanged, new_value)
```
```suggestion
    valueChanged: during hwobj initialization and when setting a new value. Format: (valueChanged, new_value)
```

axelboc commented on mxcubecore/HardwareObjects/abstract/AbstractActuator.py:
```diff
@@ -37,7 +44,21 @@
+        actuator_name (str):
+            actuator name specified by XML property, otherwise `None`
+        read_only (bool):
+            read only flag, specified by XML property, otherwise `False`
```
```suggestion
            read-only flag, specified by XML property, otherwise `False`
```

axelboc commented on mxcubecore/HardwareObjects/abstract/AbstractActuator.py:
```diff
@@ -175,10 +204,10 @@ def re_emit_values(self):
     def force_emit_signals(self):
-        """Forces to emit all signals.
+        """
+        Forces to emit all signals.
```
"Force" seems implied.

```suggestion
        Emit all signals.
```

Alternatively, in the next file, you've used the word "Reemit".

axelboc commented on mxcubecore/HardwareObjects/abstract/AbstractAperture.py:
```diff
@@ -53,21 +53,24 @@ def init(self):
 
     def get_diameter_size_list(self):
-        """
+        """Get diameters size list
```
Might as well be a bit more verbose:

```suggestion
        """Get list of diameter sizes.
```

axelboc commented on mxcubecore/HardwareObjects/abstract/AbstractAperture.py:
```diff
@@ -53,21 +53,24 @@ def init(self):
 
     def get_position_list(self):
-        """
+        """Get positions list
```
```suggestion
        """Get list of positions.
```

and so on ...

It's not always needed; for instance, the next one reads fine: "Get current diameter index"

axelboc commented on mxcubecore/HardwareObjects/abstract/AbstractAperture.py:
```diff
@@ -171,15 +181,20 @@ def set_out(self):
     def force_emit_signals(self):
-        """
-        Reemits all signals
+        """Reemits all signals
```
I would stick with the indicative mood everywhere:

```suggestion
        """Reemit all signals.
```

axelboc commented on mxcubecore/HardwareObjects/abstract/AbstractCharacterisation.py:
```diff
@@ -151,7 +155,8 @@ def get_default_characterisation_parameters(self, default_input_file):
         """Called by characterise before characterisation starts, can be used
         to manipulate edna_input object before characterisation starts.
-        Example: to set a site specific output directory
+        Example: to set a site specific output directory.
```
```suggestion
        Example: to set a site-specific output directory.
```

axelboc commented on mxcubecore/HardwareObjects/abstract/AbstractCollect.py:
```diff
@@ -143,7 +148,15 @@ def collect(self, owner, dc_parameters_list):
     def do_collect(self, owner):
         """
-        Actual collect sequence
+        Actual collect sequence.
```
```suggestion
        Actually collect the sequence.
```

axelboc commented on mxcubecore/HardwareObjects/abstract/AbstractCollect.py:
```diff
@@ -317,7 +340,12 @@ def collection_stopped(self):
 
     def collection_finished(self):
-        """Collection finished beahviour"""
+        """Collection finished beahviour
```
```suggestion
        """Collection finished behaviour
```

axelboc commented on mxcubecore/HardwareObjects/abstract/AbstractCollect.py:
```diff
@@ -747,14 +834,16 @@ def move_to_centered_position(self):
     def take_crystal_snapshots(self):
         """
-        Descript. :
+        Take crytal snapshots of the currently loaded sample
```
```suggestion
        Take crystal snapshots of the currently loaded sample
```

meguiraun committed:
Apply suggestions from code review

Co-authored-by: Axel Bocciarelli <axelboc@gmail.com>

meguiraun committed:
Update mxcubecore/HardwareObjects/abstract/AbstractCharacterisation.py

Co-authored-by: Axel Bocciarelli <axelboc@gmail.com>

meguiraun committed:
text styling

beteva reviewed (commented):
Very nice.
Apparently we need changes in many python files for better documentation. I suggest we see together with the other developers who does what, but I like the class description you've made for AbstractActuator. We should follow the same way.

meguiraun committed:
documenting emits at class level

meguiraun commented:
AbstractBeam with docstrings in class [here](https://mxcubecore--866.org.readthedocs.build/en/866/dev/autosummary/mxcubecore.HardwareObjects.abstract.AbstractBeam.html)
AbstracActuator with docstrings outside class [here](https://mxcubecore--866.org.readthedocs.build/en/866/dev/autosummary/mxcubecore.HardwareObjects.abstract.AbstractActuator.html)

I can remove all changes to the abstract classes if it makes the other PR easier to handle

ping @beteva

fabcor-maxiv commented:
> AbstractBeam with docstrings in class [here](https://mxcubecore--866.org.readthedocs.build/en/866/dev/autosummary/mxcubecore.HardwareObjects.abstract.AbstractBeam.html)
> AbstracActuator with docstrings outside class [here](https://mxcubecore--866.org.readthedocs.build/en/866/dev/autosummary/mxcubecore.HardwareObjects.abstract.AbstractActuator.html)

For me, this belongs in the class docstring : )

meguiraun commented:
> > AbstractBeam with docstrings in class [here](https://mxcubecore--866.org.readthedocs.build/en/866/dev/autosummary/mxcubecore.HardwareObjects.abstract.AbstractBeam.html)
> > AbstracActuator with docstrings outside class [here](https://mxcubecore--866.org.readthedocs.build/en/866/dev/autosummary/mxcubecore.HardwareObjects.abstract.AbstractActuator.html)
> 
> For me, this belongs in the class docstring : )

I think same way, it seems its place

beteva commented:
@meguiraun, agree the best place is the class docstrings.

marcus-oscarsson commented:
Hey @meguiraun, do you have an update on this one, it seemed almost done ?

meguiraun commented:
completely lost track of this one... so sorry...

If people are happy I am also happy. Not all the hwobjs are covered tough

fabcor-maxiv commented:
Not all checks are successful. After this is fixed, I think it is fine to merge. And then I might do follow-ups as I had already started in <https://github.com/mxcube/mxcubecore/pull/947>.

meguiraun committed:
linting

meguiraun committed:
linting

meguiraun commented:
all checks are green now!

marcus-oscarsson committed:
[skip ci] Bumped minor version

marcus-oscarsson commented:
Sorry my mistake. I must have missed the last comments, was this ready to be merged ?

marcus-oscarsson committed:
Merge branch 'develop' into signal_documentation

marcus-oscarsson commented:
@fabcor-maxiv, @elmjag. Sorry to take such a long time with this one. Is this one ready to be merged now. I think there's been a conflict for some time now ?

fabcor-maxiv commented:
It is still on my mind. I plan to take over this one and also integrate some of the work I started in <https://github.com/mxcube/mxcubecore/pull/947>. I assigned it to me.

fabcor-maxiv commented:
Done in <https://github.com/mxcube/mxcubecore/pull/1116>.
